### PR TITLE
Adjust chunksize if necessary

### PR DESCRIPTION
**Pull request by JordonPhillips**
This will adjust the chunksize of an upload or copy if it is too small, too
large, or if it looks like max parts will be exceeded. This is unnecessary
for downloads since there are no such restrictions for ranged downloading.

cc @kyleknap @jamesls

### CODE REVIEW

**Comment by jamesls** [comment]
What do you all think about not making the adjust chunk size "global"?  That's one of the quirks I remember from the CLI/boto3 that this meant you had to update every single test and use "realistic" sized data in unit tests when they don't really matter (for those particular tests).

What if these objects instead took a "part size adjuster".  We could then have the tests use the "noop" version for unit tests that don't actually care and in the "real" code we could pass the actual chunk size adjuster.

Also, just a heads up, the travis build failed in case you haven't seen it yet.

**Comment by jamesls** [comment]
Just wanted to capture offline conversation I had with @kyleknap .  Let's go ahead and update this to be configurable.  I think having a this validation be global was reasonable for the CLI, but I think for this library it makes sense to allow this to be configurable.

@kyleknap let me know if I missed anything.

**Comment by JordonPhillips** [comment]
I could see adding this to our transfer config or as an optional parameter to `copy` and `upload`, but it seems odd to change our public interface to make testing marginally simpler.

**Comment by JordonPhillips** [comment]
As we discussed offline, there's some consideration on how we would want to open this up in the future. In the meantime, we'll settle for leaving it open to configuration by allowing an adjuster to be provided with `main_kwargs`.

File: s3transfer/utils.py [hunk 1]
```diff
@@ -567,3 +574,57 @@ def release(self, tag, acquire_token):
...
+
+
+class ChunksizeAdjuster(object):
+    def __init__(self, max_size=MAX_SINGLE_UPLOAD_SIZE,
+                 min_size=MIN_UPLOAD_CHUNKSIZE, max_parts=MAX_PARTS):
+        self.max_size = max_size
+        self.min_size = min_size
+        self.max_parts = max_parts
+
+    def adjust_chunksize(self, current_chunksize, file_size=None):
```
Comment (jamesls) [inline s3transfer/utils.py]: Why do we allow file_size to be `None`?  Do we actually need that behavior?

**Comment by jamesls** [comment]
I'll probably defer to @kyleknap on this, but after seeing this version, I'm a little hesitant about this new concept of default values in task submission.  These are essentially params that will never be invoked by "real" code (i.e through the TransferManager public API).

If we're not going to offer a public API right now, I'm not sure if there's much benefit in plumbing this in right now.

I suppose that puts me at -0 but I can be convinced.

File: s3transfer/utils.py (hunk 1 above)
Comment (JordonPhillips) [inline s3transfer/utils.py]: Yes. We don't know the file size when uploading non-seekable files.

Comment (kyleknap) [inline s3transfer/utils.py]: That makes sense. It would be good to document what ``None`` implies here though.

File: tests/unit/test_utils.py [hunk 2]
```diff
@@ -683,3 +685,34 @@ def acquire():
...
+        new_size = self.adjuster.adjust_chunksize(chunksize, file_size)
+        self.assertEqual(new_size, MAX_SINGLE_UPLOAD_SIZE)
+
+    def test_chunksize_too_small(self):
+        chunksize = 7 * (1024 ** 2)
+        file_size = 5 * (1024 ** 4)
+        # If we try to upload a 5TB file, we'll need to use 896MB part
+        # sizes.
+        new_size = self.adjuster.adjust_chunksize(chunksize, file_size)
+        self.assertEqual(new_size, 896 * (1024 ** 2))
```
Comment (kyleknap) [inline tests/unit/test_utils.py]: We should also test for when the file size is ``None``.

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -683,3 +685,34 @@ def acquire():
...
+        file_size = MAX_SINGLE_UPLOAD_SIZE * 2
+        new_size = self.adjuster.adjust_chunksize(chunksize, file_size)
+        self.assertEqual(new_size, MAX_SINGLE_UPLOAD_SIZE)
+
+    def test_chunksize_too_small(self):
+        chunksize = 7 * (1024 ** 2)
+        file_size = 5 * (1024 ** 4)
+        # If we try to upload a 5TB file, we'll need to use 896MB part
+        # sizes.
+        new_size = self.adjuster.adjust_chunksize(chunksize, file_size)
```
Comment (kyleknap) [inline tests/unit/test_utils.py]: It would also be good to make sure that the number of parts you get out of it is less than the max number of parts.

**Comment by kyleknap** [comment]
Yeah I would prefer holding off on making the ``ChunksizeAdjuster`` an optional parameter you can provide to ``_submit()`` mainly because it is not really used in the code at all.

Otherwise it looks good. Just had comments on the tests.

**Comment by JordonPhillips** [comment]
Updated

File: tests/unit/test_utils.py [hunk 4]
```diff
@@ -683,3 +686,41 @@ def acquire():
...
+        self.assertEqual(new_size, 896 * (1024 ** 2))
+        num_parts = file_size / new_size
+        self.assertLessEqual(num_parts, MAX_PARTS)
+
+    def test_unknown_file_size(self):
```
Comment (jamesls) [inline tests/unit/test_utils.py]: It would be nice to add a test when we don't know the file size and we actually make an adjustment (either above/below the thresholds).

**Comment by jamesls** [comment]
Looks good to me, just had a small request for a test.

**Comment by JordonPhillips** [comment]
Tests added

**Comment by kyleknap** [comment]
Thanks. Looks good. :ship: Also make sure you add a changelog entry.

**Comment by jamesls** [comment]
:shipit:
